### PR TITLE
Add SIGTERM handling in web.run_app

### PR DESCRIPTION
**Pull request by cecton**
## What do these changes do?

Makes `web.run_app()` handle SIGTERM by default to gracefully exit the application.

## Are there changes in behavior for the user?

The application run with `web.run_app()` will override any existing SIGTERM handler in the current process.

## Related issue number

#1932 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new entry to `CHANGES.rst`
  * Choose any open position to avoid merge conflicts with other PRs.
  * Add a link to the issue you are fixing (if any) using `#issue_number` format at the end of changelog message. Use Pull Request number if there are no issues for PR or PR covers the issue only partially.

### CODE REVIEW

**Comment by cecton** [comment]
The testing is a bit complicated so I only tested using the script given in the original issue.

File: aiohttp/web.py [hunk 1]
```diff
@@ -414,12 +423,15 @@ def run_app(app, *, host=None, port=None, path=None, sock=None,
         asyncio.gather(*server_creations, loop=loop)
     )
 
+    if handle_signals:
+        signal.signal(signal.SIGTERM, sigterm_handler)
```
Comment (asvetlov) [inline aiohttp/web.py]: https://docs.python.org/3/library/asyncio-eventloop.html#unix-signals should be used for signal registration.
Also signal subscription should be skipped on Windows

File: aiohttp/web.py [hunk 2]
```diff
@@ -318,10 +319,18 @@ def __repr__(self):
...
+
+def sigterm_handler(signum, frame):
+    raise TerminateSignal()
+
+
 def run_app(app, *, host=None, port=None, path=None, sock=None,
             shutdown_timeout=60.0, ssl_context=None,
             print=print, backlog=128, access_log_format=None,
-            access_log=access_logger, loop=None):
+            access_log=access_logger, handle_signals=True, loop=None):
```
Comment (asvetlov) [inline aiohttp/web.py:333]: maybe `handle_sigterm` is better name?

Comment (cecton) [reply aiohttp/web.py:333]: Yes actually I was thinking that you may want actually that run_app handles all the signals or none. For the user it's not really about handling SIGTERM in particular, it's more about handling the signals (INT and TERM) or not. It's unfortunate somehow that Python handles SIGINT for you by default.

I just realize now that maybe we should also handle SIGINT using https://docs.python.org/3/library/asyncio-eventloop.html#unix-signals but I'm not sure at all about it, what do you think?

**Comment by cecton** [comment]
@asvetlov Do you mind if I move the print() inside the try...except? https://github.com/aio-libs/aiohttp/pull/1946/commits/87fa801fa85f5d07cec6b2d5f7673d8fae6e8c3a

Otherwise I can't possibly know when to terminate.

File: aiohttp/web.py (hunk 2 above)
Comment (asvetlov) [reply aiohttp/web.py:333]: SIGINT is handled out of the box, why should we care about it?

**Comment by asvetlov** [comment]
>  Do you mind if I move the print() inside the try...except?
I totally ok with it

**Comment by asvetlov** [comment]
Could you use asyncio api for subprocesses in test instead of `subprocess.Popen`?
It's not critical but why not if it is exist.

**Comment by cecton** [comment]
@asvetlov Yes that's what I wanted in the beginning but the test method is not asynchronous. There is no point and it would make thing unreadable.

**Comment by cecton** [comment]
@asvetlov I mean:

The test function is synchronous so I suppose the tests are run synchronously. The thing I'm testing spawns a process, wait for it and assert. There is only one thing done at a time.

If I want to make it asynchronous it means I will have to move the test to a coroutine and the test itself will do a `run_until_complete` on that coroutine.

There is no point of doing so because the event loop will be busy with only one task. And the perk of using `asyncio.create_subprocess` is also not so good but it is actually using subprocess behind the scene (it's just a wrapper).

So unless you know how to make the tests run asynchronously in parallel like mocha.parallel, there is no real advantage to this.

```python
async def real_test():
    proc = await asyncio.create_subprocess....
    async for line in proc.stdout:
        ....
    proc.terminate()
    retcode = await proc.wait()
    assert ...

def test_sigterm(loop):
    loop.run_until_complete(real_test())
```

**Comment by cecton** [comment]
@asvetlov I just checked the code of asyncio and I believe we should also manage the SIGINT the same way.

The reason is simply that if not managed, Python has this default behavior to raise KeyboardInterrupt. But apparently they made `add_signal_handler` in asyncio because using `signal.signal` wasn't enough.

This is from asyncio:

```python
    def add_signal_handler(self, sig, callback, *args):
(...)
        handle = events.Handle(callback, args, self)
        self._signal_handlers[sig] = handle

        try:
            # Register a dummy signal handler to ask Python to write the signal
            # number in the wakup file descriptor. _process_self_data() will
            # read signal numbers from this file descriptor to handle signals.
            signal.signal(sig, _sighandler_noop)

            # Set SA_RESTART to limit EINTR occurrences.
            signal.siginterrupt(sig, False)
(...)
```

This is the doc for siginterrupt:
> Change system call restart behaviour: if flag is False, system calls will be restarted when interrupted by signal signalnum, otherwise system calls will be interrupted. Returns nothing. Availability: Unix (see the man page siginterrupt(3) for further information).
>
> **Note that installing a signal handler with signal() will reset the restart behaviour to interruptible by implicitly calling siginterrupt() with a true flag value for the given signal.**
From https://docs.python.org/3/library/signal.html#signal.siginterrupt

The other reason that makes me think we should handle it is that they are managing the signals with the event loop itself.

I'm not sure what are the consequences because KeyboardInterrupt and the `signal.signal` call I just added work fine in a simple example but I have a strong feeling about it. I propose to keep the variable `handle_signal` and actually handle INT and TERM with `add_signal_handler`.

File: aiohttp/web.py (hunk 2 above)
Comment (cecton) [reply aiohttp/web.py:333]: I actually answered later in the PR comments. It's not really "out of the box" for the event loop, it's out of the box for default synchronous usage of Python.

**Comment by cecton** [comment]
@asvetlov I'm going to give up the signal testing on Windows if you don't mind...

Apparently the console ask to confirm when you send a signal CTRL_BREAK_EVENT or CTRL_C_EVENT:

```
tests\test_resolver.py ssssss....s.sss..
tests\test_run_app.py ...................sssss.s.^CTerminate batch job (Y/N)? 
```

By the way, you may have notice that your Windows CI (appveyor) is a bit stuck? This is because each build wait 1 hour before being killed. Since there are 6 builds every time, it's waiting 6 hours. I didn't realize fast enough and I'm afraid I'm blocking your CI for 7 hours now and there are still 5 hours to go.

Hum.... sorry?

**Comment by asvetlov** [comment]
Oooh. If Windows hangs with user prompt -- please don't run test on it. I'm fine with your proposal.
Don't worry about appveyor blocking -- I believe it was not intentional :)

**Comment by cecton** [comment]
@asvetlov ok finally done :) Are you sure about the subprocess? Did I convince you?

**Comment by asvetlov** [comment]
Thanks

**Pull request merged**